src/pretrain_utils.py

```diff
@@ -97,99 +101,103 @@ def make_masked_patch_prediction_target(patch_tokens):
+    mask_position = tf.cast(mask_position, tf.int32)
+    num_real_masked_tokens = tf.reduce_sum(mask_position, axis=1, keepdims=True)
+    position = tf.range(masked_seq_len, dtype=tf.int32)
+    return tf.where(position < num_real_masked_tokens, 1.0, 0.0)
 
   def make_masked_language_model_and_masked_patch_prediction_features(
     features: Mapping[str, tf.Tensor]) -> Mapping[str, tf.Tensor]:
     """Randomly mask out text and patch tokens
     Args:
       features: A dictionary of Tensor features, crucially including
-          `image_input_ids`, `text_input_ids`.
+          `patch_input_ids`, `text_input_ids`.
     Returns:
       A new `features` dictionary with global-local transformer side inputs.
     """
 
     features = dict(features)
 
-    # Image: create features for masked patch prediction (mpp)
-    # image_input_ids: <tf.RaggedTensor>[batch, (patches), (1)].
-    (image_masked_token_ids,
-     image_masked_positions, image_masked_ids) = tf_text.mask_language_model(
-        features['image_input_ids'],
-        image_item_selector,
+    # Patch: create features for masked patch prediction (mpp)
+    # patch_input_ids: <tf.RaggedTensor>[batch, (patches), (1)].
+    (masked_patch_token_ids,
+     masked_patch_positions, _) = tf_text.mask_language_model(
+        features.pop('patch_input_ids'),
+        patch_item_selector,
         mask_values_chooser,
         axis=1)
-
+        
     # Offset -2 position indices: [CLS] and [PATCH] tokens.
-    shifted_image_masked_positions = (image_masked_positions - 2).to_tensor()
-    batch_size, image_masked_seq_len = tf_utils.get_shape_list(
-        shifted_image_masked_positions)
-    shifted_image_masked_positions = tf.cast(shifted_image_masked_positions,
+    # We want to select from patch embeddings.
+    shifted_patch_masked_positions = (masked_patch_positions - 2).to_tensor()
+    shifted_patch_masked_positions = tf.cast(shifted_patch_masked_positions,
                                              tf.int32)
-    patch_token_size = tf_utils.get_shape_list(features['image_data'])[2]
-    masked_patch_tensor = gather_indexes(features['image_data'],
-                                         shifted_image_masked_positions)
-    masked_patch_tensor = tf.reshape(masked_patch_tensor,
-                                     [batch_size,
-                                      image_masked_seq_len,
-                                      patch_token_size])
-    
-    features['image_masked_positions'] = tf.cast(image_masked_positions,
-                                                 tf.int32).to_tensor()
+    batch_size = tf_utils.get_shape_list(features['patch_embeddings'])[0]
+    masked_patch_embeddings = tensor_utils.gather_indexes(
+      features['patch_embeddings'], shifted_patch_masked_positions)
+    masked_patch_embeddings = rearrange(masked_patch_embeddings,
+                                        '(b s) h -> b s h',
+                                        b=batch_size)
```

Comment:
What does this rearrange do?

Reply:
This rearrange is to disentangle the first two dimension (batch_size and seq_len). It's because these two dimension will be merged after gathering by indices.